starlette/templating.py

```diff
@@ -23,20 +24,26 @@ class _TemplateResponse(Response):
     def __init__(
         self,
         template: typing.Any,
-        context: dict,
+        context: dict = None,
         status_code: int = 200,
         headers: dict = None,
         media_type: str = None,
         background: BackgroundTask = None,
     ):
         self.template = template
-        self.context = context
-        content = template.render(context)
-        super().__init__(content, status_code, headers, media_type, background)
+        self.context = context or {}
+        super().__init__(b"", status_code, headers, media_type, background)
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if "request" not in self.context:
+            self.context["request"] = Request(scope, receive, send)
```

Comment:
Not really sure it would be out of scope for this PR or not.
But now that the request is set internally, I propose to rename it to `_request` so it doesn't interfere with user's `context`.

Reply:
I'm not sure I understand the motivation of having `"_request"` in the template context automatically.

Using...

```python
        if "request" not in self.context:
            self.context["request"] = Request(scope, receive, send)
```

seems pretty sensible to me. Can you talk me through the motivation of changing that?

Reply:
If we do something like:

```python
@app.route("/")
    async def homepage(request):
        return templates.TemplateResponse("index.html", {"request": "example"})
```

Then the `url_for` will throw and exception as it's always expecting the `request` to be Starlette  `Request` object. We're somehow making the `request` keyword unavailable to template context.
Does that make sense?

Reply:
I'm not entirely sure we still need the `if` check anymore. If we set it internally, I can't think of any reason why the user request would be different from the one created by TemplateResponse.

Reply:
I think a reasonable API would be that we document that `TemplateResponse` includes `"request"` by default.

It's such a semantically obvious name, that I think a user including `"request"` and setting it to something *other* that a `Request` instance is causing trouble for themselves.

Reply:
Great, thanks for the feedback.